runners/direct-java/src/main/java/org/apache/beam/runners/direct/GroupByKeyOnlyEvaluatorFactory.java

```diff
@@ -114,16 +116,31 @@ public void processElement(WindowedValue<KV<K, V>> element) {
       KV<K, V> kv = element.getValue();
       K key = kv.getKey();
       StructuralKey<K> groupingKey = StructuralKey.of(key, keyCoder);
-      List<WindowedValue<V>> values =
-          groupingMap.computeIfAbsent(groupingKey, k -> new ArrayList<>());
+      Map<StructuralKey<K>, List<WindowedValue<V>>> curMap;
+      if (element.isRetraction()) {
+        curMap = retractionGroupingMap;
+      } else {
+        curMap = groupingMap;
+      }
+
+      List<WindowedValue<V>> values = curMap.computeIfAbsent(groupingKey, k -> new ArrayList<>());
       values.add(element.withValue(kv.getValue()));
     }
 
     @Override
     public TransformResult<KV<K, V>> finishBundle() {
       StepTransformResult.Builder resultBuilder = StepTransformResult.withoutHold(application);
-      for (Map.Entry<StructuralKey<K>, List<WindowedValue<V>>> groupedEntry :
-          groupingMap.entrySet()) {
+      processGroupedValues(resultBuilder, retractionGroupingMap, true);
```

Comment:
This is an interesting case. Does this transform obey the specification?

`resolveRetractions(transform(pc)) == resolveRetractions(transform(resolveRetractions(pc)))`

I think your version might. That is nice. It is not required, since this is an internal transform, but it will make things much easier for everyone.

Another approach is to make `KeyedWorkItem` able to hold additions and deletions. Then it will not obey the property above. But a `KeyedWorkItem` is really an intermediate hack to store additions, deletions, and timers, to feed them to GroupAlsoByWindow.

And another approach, which is used by #4742, is to do nothing. It already works as-is. The metadata about which elements are retractions and which are not is already copied by `element.withValuie(kv.getValue())` on line 127. So with no changes it already outputs a list where each item is labeled as addition or retraction.